posts/2021/09/array-libraries-interoperability.md

```diff
@@ -0,0 +1,537 @@
+	when used together. But it ain't about finding a way to use them
+	individually; it's about making them work together.
+	That's how interoperable science is done." ~ ([actual quote :P](https://youtu.be/8xFEqdkO5UI?t=13))
+
+OK sorry, that's just me using Rocky's motivational lines to make a point.
+To define the issue more concretely, the question is, can we do something like
+the following?!
+
+```python
+import torch
+
+...  # Do something with PyTorch
+x = torch.rand(3000)  # End up with some torch tensor
+
+# Estimate power spectral density using Welch’s method.
+# SciPy offers a method `welch` for doing exactly that.
+from scipy.signal import welch
+
+f, Pxx = welch(x, fs=10e3, nperseg=1024)
+```
+
+Our world would be a better place if we could pass any kind of "array provider"
+object in these consumer libraries and let them do their magic, finally
+spitting out the same "array provider" type as the one in input. Note: All the
+arguments should be of the same kind.
+
+What if there is a way? What if there are existing ways to achieve this?
+Not kidding, it is possible with recent efforts made towards this direction.
+
+## Protocols for Interoperable Behaviour
+
+Now that the problem and motivation is pellucid, let's dive into the
+technicalities involved in array libraries interoperability and 
+understand the protocols making this a reality.
+
+Enter [NEP 18 (`__array_function__` protocol)](https://numpy.org/neps/nep-0018-array-function-protocol.html), which was one of the first
+efforts to address the issue of interoperability in NumPy. In a nut shell
+NEP 18 allows arguments of NumPy functions to define how that function
+operates on them. This enables using NumPy as a high level API for
+efficient multi-dimensional array operations, even with array implementations
+that differ greatly from `numpy.ndarray`.
+
+I suggest reading the [NEP](https://numpy.org/neps/nep-0018-array-function-protocol.html)
+itself for detailed understanding, but I'll try to expound the motivation
+with a simple example taken from [this](https://www.youtube.com/watch?v=HVLPJnvInzM)
+insightful talk by [Dr. Ralf Gommers](https://github.com/rgommers)
+at PyData Amsterdam.
+
+
+```python
+import numpy as np
+import cupy
+
+def some_func(x):
+    return np.log(np.sin(x) + 5)
+
+x = np.random.random((100000, 100000))
+
+some_func(x)  # Runs on CPU, might be slow
+
+# Now can we use some_func with CuPy arrays
+# designed to work on NVIDIA GPUs and are
+# faster at parallelized matrix operations.
+
+x_cupy = cupy.array(x)
+
+# NEP 18 enables this possibility
+some_func(x_cupy) # Runs on GPU, orders of magnitude fast
+```
+
+Since NEP 18, there have been a few other protocols like
+[NEP 35](https://numpy.org/neps/nep-0035-array-creation-dispatch-with-array-function.html),
+[NEP 37](https://numpy.org/neps/nep-0037-array-module.html) endeavouring to
+address some of the issues and shortcomings with NEP 18.
```

Comment:
would be good to point out that these NEPs were not actually accepted/implemented